### ISSUE TITLE
[BUG] Sentence end defined incorrectly

### ISSUE THREAD

**Issue by ftonneau** [mawww/kakoune#5134]
### Version of Kakoune

Development version or current version on Arch

### Reproducer

Write this in an empty buffer:

`My name is A.B. Jones. Be my guest.
`

Position your cursor at line start (on "M"), then select an outer sentence with `<a-a>s`
 

### Outcome

Kakoune selects "My name is A."

### Expectations

Kakoune should select "My name is A.B. Jones. "

### Additional information

From `selectors.cc`, Kakoune defines the end of a sentence as one of `.;!?` characters. This is incorrect in English as well as other Western languages. The end of a sentence is better defined as one of `.;!?` characters followed by one or two horizontal spaces or a line return. (A few corner cases could also be considered in English, as when the ending colon is followed by a closing quote, but including space after `.;!?` would at least take care of the most common cases.)

**Comment by ftonneau** [comment]
Of course, the example should be:

`My name is A.B.Jones. Be my guest.` 

(facepalm). I mentioned the lack of space requirement after `.;!?` on the Kakoune forum years ago, but never filled the bug report.

**Comment by ftonneau** [comment]
A better example (for real):

`Philosophers (e.g., Fodor, 1975) and linguists (e.g., Chomsky, 1959) disagree.`

Placing the cursor on `P` and extending to sentence end repeatedly results in 4 false stops (on `e.` ,`g.`, `e.`, `g.`) And we cannot even repeat the last object selection directly (with `<a-.>`), because at each stage we are stuck on the colon. Instead, at each stage we need to extend the selection to the right a little bit before typing `<a-.>` and getting unstuck. Kakoune's support for sentence ending should definitely be improved.

**Comment by mawww** [comment]
I agree this is something to be fixed, I'll try to dedicate a bit of time to that.

**Comment by schragge** [comment]
FYI, this is how [sentence is defined](https://vimdoc.sourceforge.net/htmldoc/motion.html#sentence) in Vim's help (`:h sentence`):
> A sentence is defined as ending at a '.', '!' or '?' followed by either the end of a line, or by a space or [tab](https://vimdoc.sourceforge.net/htmldoc/intro.html#tab).  Any number of closing ')', ']', '"' and ''' characters may appear after the '.', '!' or '?' before the spaces, tabs or end of line.  A [paragraph](https://vimdoc.sourceforge.net/htmldoc/motion.html#paragraph) and [section](https://vimdoc.sourceforge.net/htmldoc/motion.html#section) boundary is also a sentence boundary.
> If the 'J' flag is present in ['cpoptions'](https://vimdoc.sourceforge.net/htmldoc/options.html#'cpoptions'), at least two spaces have to follow the punctuation mark; `<Tab>`s are not recognized as white space.
> The definition of a sentence cannot be changed.

This logic is implemented by Vim in function [findsent](https://github.com/vim/vim/blob/master/src/textobject.c#L18).

**Comment by ftonneau** [comment]
My opening example was completely and stupidly messed up. My latter example is a better one:

`Philosophers (e.g., Fodor, 1975) and linguists (e.g., Chomsky, 1959) disagree.`

Here Kakoune will detect a sentence end at five different places, the first four ones being false positives because they involve a colon not followed by a space.

It is true that no reasonable definition will eliminate all false positives (e.g., the colon in `Dr. Jones`), but a definition such as Vim's is certainly much better than Kakoune's because contrary to the latter, Vim's definition eliminates more false positives.

**Comment by kamurani** [comment]
@ftonneau  just an FYI, a colon is a `:` character, and `.` is called a "period" or "full-stop" 
Got a bit confused reading your comments.

**Comment by ftonneau** [comment]
You are right, thanks for correcting.
